### PR TITLE
CA-289145: close socket if error occurs when using lwt connect

### PR DESCRIPTION
**Pull request by krizex**
Signed-off-by: Yang Qian <yang.qian@citrix.com>

### CODE REVIEW

**Review by mseri — CHANGES_REQUESTED** [review]
In all those I think you need to use `Lwt.fail` for the exceptions,  `Lwt.catch` instead of `try...with` and bind the close: `(fun e -> Lwt_unix.close sock >>= fun () -> Lwt.fail e)`. Or something like that

**Comment by krizex** [comment]
Thanks @mseri , updated.

File: src/newcli.ml [hunk 1]
```diff
@@ -253,7 +263,12 @@ let open_channels () =
     try_lwt
```
Comment (mseri) [inline src/newcli.ml:263]: This is the old camlp4 syntax... this makes me wonder if this package is maintained at all. Does it even compile in the new xs-opam?

**Comment by mseri** [comment]
Sorry for not spotting this earlier, this repository needs to be deprecated and archived as it is no longer maintained and the new xe is shipped and buildable from the main xen-api repository

**Comment by mseri** [comment]
@robhoes or @jonludlam can you archive this?

File: src/newcli.ml [hunk 2]
```diff
@@ -234,7 +234,12 @@ let open_tcp_ssl host =
   Lwt_unix.gethostbyname host >>= fun host_entry ->
   let sockaddr = Lwt_unix.ADDR_INET(host_entry.Lwt_unix.h_addr_list.(0), get_xapiport true) in
   let sock = socket sockaddr in
-  Lwt_unix.connect sock sockaddr >>= fun () ->
+  Lwt.catch (fun () ->
+      Lwt_unix.connect sock sockaddr
+    ) (fun e ->
+      Lwt_unix.close sock >>= Lwt.fail e
```
Comment (mseri) [inline src/newcli.ml:240]: `>>=` should be `>>= fun () ->`

File: src/newcli.ml [hunk 3]
```diff
@@ -244,7 +249,12 @@ let open_tcp host =
     Lwt_unix.gethostbyname host >>= fun host_entry ->
     let sockaddr = Lwt_unix.ADDR_INET(host_entry.Lwt_unix.h_addr_list.(0), get_xapiport false) in
     let sock = socket sockaddr in
-    Lwt_unix.connect sock sockaddr >>= fun () ->
+    Lwt.catch (fun () ->
+        Lwt_unix.connect sock sockaddr
+      ) (fun e ->
+        Lwt_unix.close sock >>= Lwt.fail e
```
Comment (mseri) [inline src/newcli.ml:255]: `>>=` should be `>>= fun () ->`

**Pull request closed without merging**